crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -460,12 +460,41 @@ def check_schema_migrations
         false
       end
 
+      # Check the version of installed SLES to verify admin upgrade went well.
+      # We have to do the check here at the start of 'services' step and not during admin upgrade,
+      # because admin upgrade step is not repeatable.
+      def check_product_version
+        # "zypper products -i" should look like: (shortened from real output)
+        # S  | Repository         | Internal Name   | Name                                | Version
+        # ---+--------------------+-----------------+-------------------------------------+--------
+        # i  | Cloud              | suse-openstack..| SUSE OpenStack Cloud Crowbar 9      | 9-0
+        # i+ | SLES12-SP4-Updates | SLES            | SUSE Linux Enterprise Server 12 SP4 | 12.4-0
+        cmd = run_cmd(
+          "LANG=C zypper products -i | grep SLES | tr -d ' ' | cut -d '|' -f 5"
```

Comment:
why not reuse product info extraction logic from `adminrepocheck`?

Reply:
Hm, I'm not sure if it's easily extractable this time

Reply:
I can parse the xml output as well. But I do not see an easy option to reuse existing method.
I guess I should do the `zypper-retry` call, that's a good one

Reply:
ok, if you don't want to do the extra checks for "system locked" or "prompts" then it all boils down to around two lines of xml processing code to be copied... imo still better than the bash magic :o)

Reply:
Yes, I will try that.
(Although I would see bash magic as not as bad compared to xml parsing...)

Reply:
If the xml parsing was to be done in bash then I'd probably go with grep on text but in ruby with `Hash.from_xml()` at hand I think it's better to use that.